xdsl/traits.py

```diff
@@ -38,26 +58,31 @@ class Pure(OpTrait):
 
 
 @dataclass(frozen=True)
-class HasParent(OpTrait):
+class HasParent(LazyOpTrait):
     """Constraint the operation to have a specific parent operation."""
 
     parameters: tuple[type[Operation], ...]
 
-    def __init__(self, *parameters: type[Operation]):
-        if not parameters:
-            raise ValueError("parameters must not be empty")
+    def __init__(self, parameters: Callable[[], tuple[type[Operation], ...]]):
         super().__init__(parameters)
 
     def verify(self, op: Operation) -> None:
         parent = op.parent_op()
-        if isinstance(parent, self.parameters):
-            return
-        if len(self.parameters) == 1:
-            raise VerifyException(
-                f"'{op.name}' expects parent op '{self.parameters[0].name}'"
-            )
-        names = ", ".join(f"'{p.name}'" for p in self.parameters)
-        raise VerifyException(f"'{op.name}' expects parent op to be one of {names}")
+        self.resolve()
+        match self.parameters:
+            case []:
+                raise ValueError("parameters should not be empty")
```

Comment:
This should still be in the init, I think

Reply:
I think it defeat the purpose of laziness?